### PR TITLE
Add GPIO pt2: The Featuration

### PR DESCRIPTION
**Pull request by twoshark**
Peers updates, yaml config loads, and GPIO pins will run their rpio init if you uncomment those lines

Haven't run on a pi with gpio yet and a fast follow should be a gpio interface{} so we can stub in a mock gpio class for off pi development

### CODE REVIEW

File: environment/Dockerfile [hunk 1]
```diff
@@ -2,8 +2,10 @@ FROM golang:1.13.5
...
+RUN export GOFIRE_MASTER_HOST=`/sbin/ip route|awk '/default/ { print $3 }'` && export GOFIRE_MASTER=true
+
 ADD . /go/src/GoFire
 
-WORKDIR /go/src/GoFire
+WORKDIR /go/src/GoFire/src
 
-RUN go get -v
\ No newline at end of file
+RUN ls
```
Comment (dtp263) [inline environment/Dockerfile]: We can get rid of this

File: src/cluster/cluster.go [hunk 2]
```diff
@@ -60,18 +60,33 @@ func (c *Cluster) UpdatePeers(urlPath string, message PeerUpdateMessage, exclude
...
 }
 
+//NewDevice -
+func (c *Cluster) NewDevice(host string, port string) (device.Device, error) {
+	dvc := device.Device{
+		ID:   c.generateUniqueID(),
+		Host: host,
+		Port: port,
+	}
+	err := dvc.LoadSolenoids()
```
Comment (dtp263) [inline src/cluster/cluster.go]: This area is probably going to end up having more functions like 'LoadSerial' or 'LoadXIOComponent'. do you think we should pull this into a 'LoadComponents' type function now? or wait till we need it?

File: src/component/solenoid.go [hunk 3]
```diff
@@ -1,26 +1,103 @@
...
+	"time"
+)
+
 // Solenoid - base component + solenoid specific metadata
 type Solenoid struct {
-	BaseComponent
-	Voltage int
-	Type    SolenoidType
+	BaseComponent `yaml:",inline"`
+	Voltage       int           `yaml:"voltage"`
```
Comment (dtp263) [inline src/component/solenoid.go]: As discussed in person. Lets move all of the non functional descriptive data into a metadata object. This includes Voltage as that is more of a nice to have thing on the frontend side. NON-BLOCKING

File: src/component/solenoid.go [hunk 4]
```diff
@@ -1,26 +1,103 @@
...
+)
+
+//SolenoidModes -
+type SolenoidModes string
+
+const (
+	//Supply - tank supply, pilot supply and transport solenoids
+	Supply = "supply"
+	//Igniter - glowfly or other HSI
+	Igniter = "igniter"
```
Comment (dtp263) [inline src/component/solenoid.go]: not a solenoid

File: src/device/device.go [hunk 5]
```diff
@@ -17,8 +23,49 @@ func (device *Device) String() string {
 	return "Device:{ deviceId:" + strconv.Itoa(device.ID) + ", Host:" + device.Host + ", port:" + device.Port + " }"
 }
 
+//Init - Initialize all the components on this device
+func (device *Device) Init() error {
```
Comment (dtp263) [inline src/device/device.go]: You suggest you are going to return an error here but only ever return nil

File: src/device/device.go [hunk 6]
```diff
@@ -17,8 +23,49 @@ func (device *Device) String() string {
...
+	err = yaml.Unmarshal(yamlFile, &device)
+	if err != nil {
+		log.Fatalf("Unmarshal: %v", err)
+		return err
+	}
+	log.Println("Loaded the Following Solenoids: ", device.solenoidsString())
+	log.Print("Performing Initializations: ")
+	err = device.Init()
+	if err != nil {
+		return err
```
Comment (dtp263) [inline src/device/device.go]: If you are going to log.Printf errors for the other error checks maybe we should put them here

File: src/component/solenoid.go (hunk 3 above)
Comment (twoshark) [reply src/component/solenoid.go]: Roger roger

File: src/component/solenoid.go (hunk 4 above)
Comment (twoshark) [reply src/component/solenoid.go]: Removed, and igniter component stub added

File: environment/Dockerfile (hunk 1 above)
Comment (twoshark) [reply environment/Dockerfile]: Aye, this is still wip and that's a debug statement

File: src/cluster/cluster.go (hunk 2 above)
Comment (twoshark) [reply src/cluster/cluster.go]: I just stubbed in igniters and added a components interface so they can be listed together. I think we can wait a moment and I'll update this name int the next PR

File: src/device/device.go (hunk 5 above)
Comment (twoshark) [reply src/device/device.go]: the err return was missing, added

File: src/device/device.go (hunk 6 above)
Comment (twoshark) [reply src/device/device.go]: added missing print

File: src/component/solenoid.go [hunk 7]
```diff
@@ -1,26 +1,109 @@
...
+}
 
+func (s *Solenoid) healthy() bool {
+	return s.Enabled && !s.GPIO.Failed
 }
 
-// SolenoidType -
-type SolenoidType int
+// SolenoidTypes -
+type SolenoidTypes string
```
Comment (dtp263) [inline src/component/solenoid.go]: singular

File: src/main.go [hunk 8]
```diff
@@ -58,5 +59,10 @@ func configureEnvironment() {
 	viper.SetDefault("GOFIRE_PORT", 8001)
 	viper.SetDefault("GOFIRE_MASTER_PORT", 8000)
 	viper.SetDefault("GOFIRE_MASTER_HOST", "127.0.0.1")
+	viper.SetDefault("GOFIRE_MOCK_GPIO", true)
 	viper.SetDefault("CLUSTER_NAME", "MasterOfHot")
 }
+
+func getDeviceConfiguration() error {
```
Comment (dtp263) [inline src/main.go]: sup wit dis?

File: src/component/solenoid.go [hunk 9]
```diff
@@ -1,26 +1,109 @@
...
+	"time"
+)
+
 // Solenoid - base component + solenoid specific metadata
 type Solenoid struct {
-	BaseComponent
-	Voltage int
-	Type    SolenoidType
+	BaseComponent `yaml:",inline"`
+	Voltage       int           `yaml:"voltage"`
```
[the remark this answers is not in the document]
Comment (twoshark) [reply src/component/solenoid.go]: removed

File: src/component/solenoid.go (hunk 7 above)
Comment (twoshark) [reply src/component/solenoid.go]: Singularized

File: src/main.go (hunk 8 above)
Comment (twoshark) [reply src/main.go]: Good question! Deleted.

File: src/cluster/cluster.go [hunk 10]
```diff
@@ -204,7 +217,7 @@ func (c *Cluster) PrintClusterInfo() {
 	log.Println()
 
 	for i := 0; i < len(c.SlaveDevices); i++ {
-		log.Println("----Device---")
+		log.Println("----Peers---")
```
[the remark this answers is not in the document]
Comment (twoshark) [reply src/cluster/cluster.go]: fixed

**Pull request merged**